x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -69,49 +79,127 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
-                    validNodeIds.add(node.node().getId());
-                }
+                allNodeIds.add(node.node().getId());
             }
-            // Shuffle the list of nodes so the one we pick is random
-            Randomness.shuffle(validNodeIds);
-            Optional<String> nodeId = validNodeIds.stream().findAny();
+            NodesStatsRequest nodesStatsRequest = new NodesStatsRequest(allNodeIds.toArray(new String[0])).clear().
+                addMetric(NodesStatsRequest.Metric.FS.metricName()).indices(new CommonStatsFlags(CommonStatsFlags.Flag.Store));
+            getClient().admin().cluster().nodesStats(nodesStatsRequest, ActionListener.wrap((nodesStatsResponse) -> {
+                List<RoutingNode> validRoutingNodes = new ArrayList<>();
+                final Map<String, Long> nodeShardsStorageBytes = new HashMap<>();
+
+                Map<String, NodeStats> nodeStatsMap = nodesStatsResponse.getNodesMap();
+                for (String nodeId : allNodeIds) {
+                    if (nodeStatsMap.get(nodeId) != null) {
+                        List<IndexShardStats> indexShardStatsList = nodeStatsMap.get(nodeId).getIndices().
+                            getShardStats(indexMetadata.getIndex());
+                        long shardsOnCurrentNodeStorageBytes = indexShardStatsList.stream().mapToLong(indexShardStats ->
+                            Arrays.stream(indexShardStats.getShards()).mapToLong(shardStats ->
+                                shardStats.getStats().getStore().getSizeInBytes()).sum()).sum();
+                        if (shardsOnCurrentNodeStorageBytes != 0) {
+                            nodeShardsStorageBytes.put(nodeId, shardsOnCurrentNodeStorageBytes);
+                        }
+                    }
+                }
+                long indexPrimaryShardsStorageBytes = nodeShardsStorageBytes.values().stream().mapToLong(Long::longValue).sum();
+                if (indexMetadata.getNumberOfReplicas() != 0) {
+                    indexPrimaryShardsStorageBytes /= indexMetadata.getNumberOfReplicas();
+                }
+                for (RoutingNode node : routingNodes) {
+                    if (nodeStatsMap.get(node.nodeId()) != null) {
+                        FsInfo.Path totalFsInfo = nodeStatsMap.get(node.nodeId()).getFs().getTotal();
+                        long nodeTotalBytes = totalFsInfo.getTotal().getBytes();
+                        long nodeAvailableBytes = totalFsInfo.getAvailable().getBytes();
+                        long freeBytesThresholdLow;
+                        if (diskThresholdSettings.getFreeDiskThresholdLow() != 0) {
+                            freeBytesThresholdLow = (long) Math.ceil(nodeTotalBytes *
+                                diskThresholdSettings.getFreeDiskThresholdLow() * 0.01);
+                        } else {
+                            freeBytesThresholdLow = diskThresholdSettings.getFreeBytesThresholdLow().getBytes();
+                        }
```

Comment:
I'm not sure what reason you're doing this here, when constructing the `DiskThresholdSettings` the bytes are always calculated (see the `setLowWatermark(...)` call in the constructor), so I think you can always use `getFreeBytesThresholdLow()` to get the number of bytes rather than doing a calculation with the percentage?

Reply:
Gosh, so this is very confusing. @gaobinlong you're right about `getFreeBytesThresholdLow` returning `0` when the watermark is configured using percentages. Like Lee, I've also been tripped by the `[setLowWatermark](https://github.com/elastic/elasticsearch/blob/6ad3f8bfdd7c7ea1b0d8e35cba90555d9fc73250/server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java#L278)` method in `DiskThresholdSettings`.

For another PR - we should rename the methods used in `setLowWatermark` to reflect that they only maybe return something. ie. `thresholdPercentageFromWatermark` -> `thresholdPercentageFromWatermarkIfPercentageConfigured` or `maybeThresholdPercentageFromWatermark`
Similar with `thresholdBytesFromWatermark`

Reply:
Can we extract this math into a method in `DiskThresholdSettings`? ie. `getLowWatermarkAsBytes` or something named similar? (with corresponding unit tests)